### PR TITLE
Issue96 update

### LINKED ISSUE #96: Improperly implemented use of pycsw update transaction and improved logging needed
Fix:

- [x] pycsw update function 
- [ ] add logging.debug to all pycsw functions

Update error message:
```
│ [2022-05-09 14:34:50,785]                 dmci.api.app:1455 ERROR    Exception on /v1/update [POST]                                                  │
│ Traceback (most recent call last):                                                                                                                   │
│   File "/usr/lib/python3.8/site-packages/flask/app.py", line 2077, in wsgi_app                                                                       │
│     response = self.full_dispatch_request()                                                                                                          │
│   File "/usr/lib/python3.8/site-packages/flask/app.py", line 1525, in full_dispatch_request                                                          │
│     rv = self.handle_user_exception(e)                                                                                                               │
│   File "/usr/lib/python3.8/site-packages/flask/app.py", line 1523, in full_dispatch_request                                                          │
│     rv = self.dispatch_request()                                                                                                                     │
│   File "/usr/lib/python3.8/site-packages/flask/app.py", line 1509, in dispatch_request                                                               │
│     return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)                                                                     │
│   File "/usr/lib/python3.8/site-packages/dmci/api/app.py", line 67, in post_update                                                                   │
│     msg, code = self._insert_update_method_post("update", request)                                                                                   │
│   File "/usr/lib/python3.8/site-packages/dmci/api/app.py", line 126, in _insert_update_method_post                                                   │
│     err = self._distributor_wrapper(worker)                                                                                                          │
│   File "/usr/lib/python3.8/site-packages/dmci/api/app.py", line 165, in _distributor_wrapper                                                         │
│     status, valid, _, failed, skipped, failed_msg = worker.distribute()                                                                              │
│   File "/usr/lib/python3.8/site-packages/dmci/api/worker.py", line 130, in distribute                                                                │
│     obj_status, obj_msg = obj.run()                                                                                                                  │
│   File "/usr/lib/python3.8/site-packages/dmci/distributors/pycsw_dist.py", line 56, in run                                                           │
│     status, msg = self._update()                                                                                                                     │
│   File "/usr/lib/python3.8/site-packages/dmci/distributors/pycsw_dist.py", line 130, in _update                                                      │
│     ) % (self._translate(), self._worker._file_metadata_id.encode())                                                                                 │
│ AttributeError: 'UUID' object has no attribute 'encode'   
```

### PR DESCRIPTION
**Pull request by johtoblan**
**Summary**: 

Make DMCI use UUID with namespace if namespace:UUID is provided, else use UUID.
Fixes a bug in the update function caused by UUID handling
Updates CSW-interface to handle namespace + UUID

**Related issue**: 
Closes #96 

**Suggested reviewer(s)**:

**Reviewer checklist**:

* [ ] The headers of all files contain a reference to the repository license
* [ ] 100% test coverage of new code - meaning:
  * [ ] The overall test coverage increased or remained the same as before
  * [ ] Every function is accompanied with a test suite
  * [ ] Tests are both positive (testing that the function work as intended with valid data) and negative (testing that the function behaves as expected with invalid data, e.g., that correct exceptions are thrown)
  * [ ] Functions with optional arguments have separate tests for all options
* [ ] Examples are supported by doctests
* [ ] All tests are passing
* [ ] All names (e.g., files, classes, functions, variables) are explicit
* [ ] Documentation (as docstrings) is complete and understandable

The checklist is based on the S-ENDA conventions and definition of done (see [General Conventions](https://s-enda-documentation.readthedocs.io/en/latest/general_conventions.html)). The above points are not necessarily relevant to all contributions. In that case, please add a short explanation to help the reviewer.

### CODE REVIEW

**Comment by johtoblan** [comment]
Need to discuss namespace handling with @mortenwh

**Comment by johtoblan** [comment]
Need to handle search in CSW (For update and delete) with Namespace:UUID.

File: dmci/api/app.py [hunk 1]
```diff
@@ -173,6 +178,19 @@ def _distributor_wrapper(self, worker):
 
         return err
 
+    @staticmethod
+    def _check_namespace_UUID(metadata_id):
+        elements = metadata_id.split(":")
+        if len(elements) < 2:
```
Comment (johtoblan) [inline dmci/api/app.py:184]: Need to accept no-namespace

File: tests/test_api/test_app.py [hunk 2]
```diff
@@ -238,6 +239,28 @@ def testApiApp_PersistFile(tmpDir, monkeypatch):
...
 
+@pytest.mark.api
+def testApiApp_CheckNamespaceUUID():
+    testUUID = "test:7278888a-96a5-4ee5-845a-2051bb8994c8"
+    correct_UUID = uuid.UUID("7278888a-96a5-4ee5-845a-2051bb8994c8")
+
+    assert App._check_namespace_UUID(testUUID) == (correct_UUID, None)
+
+    # Test with no namespace
+    out = App._check_namespace_UUID("blabla")
```
Comment (johtoblan) [inline tests/test_api/test_app.py:250]: Should accept no namespace in delete.

**Comment by johtoblan** [comment]
PropertyIsIn is not implemented it seems, so we create namespace+:+UUID from incoming request, since we can only do exact matches with PropertyEqualsTo

**Comment by mortenwh** [comment]
See my comment in #96 - `PropertyIsEqualTo` should be used as before. Old registrations should be updated.

**Comment by johtoblan** [comment]
> See my comment in #96 - `PropertyIsEqualTo` should be used as before. Old registrations should be updated.

Yes. This is now implemented, with namespace-handling

**Review by mortenwh — APPROVED** [review]
looks good with minor edits

File: tests/test_api/test_app.py [hunk 3]
```diff
@@ -161,11 +162,12 @@ def testApiApp_InsertUpdateRequests(client, monkeypatch):
...
     assert isinstance(client, flask.testing.FlaskClient)
-    testUUID = "7278888a-96a5-4ee5-845a-2051bb8994c8"
+    testUUID = "test:7278888a-96a5-4ee5-845a-2051bb8994c8"
 
     # Invalid UUID
-    assert client.post("/v1/delete/blabla").status_code == 404
-    assert client.post("/v1/delete/7278888a96a54ee5845a2051bb8994c8").status_code == 404
+    assert client.post("/v1/delete/blabla").status_code == 400
+    # Valid UUID, but no namespace
+    assert client.post("/v1/delete/7278888a96a54ee5845a2051bb8994c8").status_code == 200
```
[the remark this answers is not in the document]
Comment (johtoblan) [reply tests/test_api/test_app.py:170]: Ja

[the remark this answers is not in the document]
Comment (mortenwh) [reply tests/test_api/test_app.py:170]: supert - kan du registrere den issue'n?

File: dmci/distributors/pycsw_dist.py [hunk 4]
```diff
@@ -127,14 +132,19 @@ def _update(self):
...
-        ) % (self._translate(), self._worker._file_metadata_id.encode())
+        ) % (self._translate(), identifier.encode(encoding="utf-8"))
         resp = requests.post(self._conf.csw_service_url, headers=headers, data=xml)
         status = self._get_transaction_status(self.TOTAL_UPDATED, resp)
 
         return status, resp.text
 
     def _delete(self):
         """Delete entry with a specified metadata_id."""
+        if self._worker._namespace != "":
```
[the remark this answers is not in the document]
Comment (johtoblan) [reply dmci/distributors/pycsw_dist.py]: See newest commit

[the remark this answers is not in the document]
Comment (johtoblan) [reply dmci/distributors/pycsw_dist.py]: and second newest. forgot to push the doc-string

File: dmci/api/app.py [hunk 5]
```diff
@@ -67,11 +67,17 @@ def post_update():
...
-        @self.route("/v1/delete/<uuid:metadata_id>", methods=["POST"])
+        @self.route("/v1/delete/<metadata_id>", methods=["POST"])
         def post_delete(metadata_id=None):
             """Process delete command."""
-            worker = Worker("delete", None, self._xsd_obj, metadata_id=metadata_id)
-            err = self._distributor_wrapper(worker)
+            metadata_uuid, namespace, err = self._check_namespace_UUID(metadata_id)
+            if metadata_uuid is not None:
+                worker = Worker("delete", None, self._xsd_obj,
+                                metadata_id=metadata_uuid, namespace=namespace)
```
Comment (mortenwh) [inline dmci/api/app.py:76]:
```suggestion
                                uuid=md_uuid, namespace=md_namespace)
```

File: dmci/api/app.py [hunk 6]
```diff
@@ -173,6 +179,25 @@ def _distributor_wrapper(self, worker):
...
+            out = uuid.UUID(elements[-1])
+            # Only UUID, no namespace
+            if len(elements) < 2:
+                return out, "", None
+            # namespace and UUID
+            elif len(elements) == 2:
+                return out, elements[0], None
+            else:
+                logger.error("Malformed metadata id. Should be <namespace>:<UUID>.")
+                return None, "", "Malformed metadata id. Should be <namespace>:<UUID>."
```
Comment (mortenwh) [inline dmci/api/app.py:195]:
```suggestion
                err = "Malformed metadata id. Should be <namespace>:<UUID>."
```

File: dmci/api/app.py [hunk 7]
```diff
@@ -173,6 +179,25 @@ def _distributor_wrapper(self, worker):
...
+        except ValueError as e:
+            logger.error(f"Failed to convert to UUID: {elements[-1]}")
+            logger.error(str(e))
+            return None, "", f"Failed to convert to UUID: {elements[-1]}"
```
Comment (mortenwh) [inline dmci/api/app.py:199]:
```suggestion
            err = f"Failed to convert to UUID: {elements[-1]}"
        
        return md_uuid, md_namespace, err
```

Comment (mortenwh) [reply dmci/api/app.py:199]: hvorfor `f"Failed to .."`?

**Review by mortenwh — CHANGES_REQUESTED** [review]
sorry - nå gikk jeg litt mye i detalj her, men med disse endringene blir det litt lettere å forstå hva som skjer...

File: dmci/api/app.py (hunk 7 above)
Comment (mortenwh) [reply dmci/api/app.py:199]: og unngå repetisjon...

**Review by mortenwh — APPROVED** [review]
Looks good!

**Pull request merged**